### PR TITLE
Add logging for all devices

### PR DESCRIPTION
**Pull request by mundruid**
- Add log utility to print error messages when exceptions occur and other information messages
- Log files are created
- Addresses issue https://github.com/networktocode/pyntc/issues/110

### CODE REVIEW

**Comment by jvanderaa** [comment]
@mundruid  This may need to get changed in the PR to using develop instead of main to help with the diffs

**Comment by mundruid** [comment]
> @mundruid This may need to get changed in the PR to using develop instead of main to help with the diffs

Thanks, @jvanderaa . I keep forgetting that the default PR in this repo is main and not develop. Changed.

**Comment by itdependsnetworks** [comment]
I am not sure if there is something unique about this logging (I see you are importing from pyntc.log as an example), but generally you are meant to use `%s` style formatting and not `f-string` formatting for logging.

**Comment by mundruid** [comment]
> I am not sure if there is something unique about this logging (I see you are importing from pyntc.log as an example), but generally you are meant to use `%s` style formatting and not `f-string` formatting for logging.

@itdependsnetworks I am happy to make the change. I was wondering why we would prefer `%` instead of `f-string`?

**Comment by itdependsnetworks** [comment]
It has to do with the lack of lazy loading in f-strings. This does a better job of explaining https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

**Comment by mundruid** [comment]
> It has to do with the lack of lazy loading in f-strings. This does a better job of explaining https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

@itdependsnetworks I have addressed the comment and replaced all the f-strings. Sorry it took so long. Let me know what you think.

File: pyntc/devices/aireos_device.py [hunk 1]
```diff
@@ -535,12 +582,20 @@ def config(self, command, **netmiko_args):
                 command_responses.append(command_response)
                 self._check_command_output_for_errors(cmd, command_response)
         except TypeError as err:
+            log.error("Host %s: Netmiko Driver's {err.args[0]}", self.host, err.args[0])
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:585]:
```suggestion
            log.error("Host %s: Netmiko Driver's %s", self.host, err.args[0])
```

File: pyntc/devices/aireos_device.py [hunk 2]
```diff
@@ -535,12 +582,20 @@ def config(self, command, **netmiko_args):
...
             raise TypeError(f"Netmiko Driver's {err.args[0]}")
         # TODO: Remove this when deprecating config_list method
         except CommandError as err:
             if not original_command_is_str:
+                log.error(
+                    "Host %s: Commands {entered_commands} returned the error {err.cli_error_msg}",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:591]:
```suggestion
                    "Host %s: Commands %s returned the error %s",
```

File: pyntc/devices/aireos_device.py [hunk 3]
```diff
@@ -535,12 +582,20 @@ def config(self, command, **netmiko_args):
...
+                    self.host,
+                    entered_commands,
+                    err.cli_error_msg,
+                )
                 raise CommandListError(entered_commands, cmd, err.cli_error_msg)
             else:
+                log.error("Host %s: Commands {entered_commands} returned the error %s", self.host, entered_commands)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:598]:
```suggestion
                log.error("Host %s: Commands %s returned the error %s", self.host, entered_commands, err.cli_error_msg)
```

File: pyntc/devices/aireos_device.py [hunk 4]
```diff
@@ -552,6 +607,12 @@ def config(self, command, **netmiko_args):
         if original_command_is_str:
             return command_responses[0]
 
+        log.info(
+            "Host %s: List of config commands %s received responses {command_response}.",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:611]:
```suggestion
            "Host %s: List of config commands %s received responses %s.",
```

File: pyntc/devices/aireos_device.py [hunk 5]
```diff
@@ -580,7 +641,7 @@ def config_list(self, commands, **netmiko_args):  # noqa: D401
             >>> device.config_list(["interface hostname virtual wlc1.site.com", "config interface vlan airway 20"])
             >>>
         """
-        warnings.warn("config_list() is deprecated; use config.", DeprecationWarning)
+        log.warning("config_list() is deprecated; use config.", DeprecationWarning)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:644]:
```suggestion
        log.warning("config_list() is deprecated; use config.")
```

File: pyntc/devices/aireos_device.py [hunk 6]
```diff
@@ -626,10 +693,12 @@ def connected(self):
         Returns:
             bool: True if the device is connected, else False.
         """
+        log.debug("Host %s: Connection status {self._connected}.", self.host, self._connected)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:696]:
```suggestion
        log.debug("Host %s: Connection status %s.", self.host, self._connected)
```

File: pyntc/devices/aireos_device.py [hunk 7]
```diff
@@ -626,10 +693,12 @@ def connected(self):
...
         return self._connected
 
     @connected.setter
     def connected(self, value):
+        log.debug("Host %s: Device connected is %s.", self.host, value)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:701]:
```suggestion
        log.debug("Host %s: Device connection status is %s.", self.host, value)
```

File: pyntc/devices/aireos_device.py [hunk 8]
```diff
@@ -673,8 +742,16 @@ def disable_wlans(self, wlan_ids):
             post_disabled_wlans = self.disabled_wlans
             if not wlans_to_validate.issubset(post_disabled_wlans):
                 desired_wlans = wlans_to_validate.union(disabled_wlans)
+                log.error(
+                    "Host %s: WLANS not disabled, with desired WLANs {desired_wlans} and post disabled WLANs {post_disabled_wlans}",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:746]:
```suggestion
                    "Host %s: Unable to disable WLANS, with desired WLANs %s and post disabled WLANs %s",
```

File: pyntc/devices/aireos_device.py [hunk 9]
```diff
@@ -673,8 +742,16 @@ def disable_wlans(self, wlan_ids):
...
+                    self.host,
+                    desired_wlans,
+                    post_disabled_wlans,
+                )
                 raise WLANDisableError(self.host, desired_wlans, post_disabled_wlans)
 
+        log.info("Host %s: All WLANs with IDs {disabled_wlans} were disabled.", self.host, disabled_wlans)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:753]:
```suggestion
        log.info("Host %s: All WLANs with IDs %s were disabled.", self.host, disabled_wlans)
```

File: pyntc/devices/aireos_device.py [hunk 10]
```diff
@@ -699,6 +776,7 @@ def disabled_wlans(self):  # noqa: D403
             >>>
         """
         disabled_wlans = [wlan_id for wlan_id, wlan_data in self.wlans.items() if wlan_data["status"] == "Disabled"]
+        log.info("Host %s: Disabled WLAN IDs: {disabled_wlans}", self.host, disabled_wlans)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:779]:
```suggestion
        log.info("Host %s: Disabled WLAN IDs: %s", self.host, disabled_wlans)
```

File: pyntc/devices/aireos_device.py [hunk 11]
```diff
@@ -757,6 +837,13 @@ def enable_wlans(self, wlan_ids):
             post_enabled_wlans = self.enabled_wlans
             if not wlans_to_validate.issubset(post_enabled_wlans):
                 desired_wlans = wlans_to_validate.union(enabled_wlans)
+                log.error(
+                    "Host %s: WLANS not enabled,\n"
+                    f"with desired WLANs {desired_wlans} and post enabled WLANs {post_enabled_wlans}",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:842]:
```suggestion
                    "with desired WLANs %s and post enabled WLANs %s",
```

File: pyntc/devices/aireos_device.py [hunk 12]
```diff
@@ -757,6 +837,13 @@ def enable_wlans(self, wlan_ids):
             post_enabled_wlans = self.enabled_wlans
             if not wlans_to_validate.issubset(post_enabled_wlans):
                 desired_wlans = wlans_to_validate.union(enabled_wlans)
+                log.error(
+                    "Host %s: WLANS not enabled,\n"
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:841]:
```suggestion
                    "Host %s: Unable to enable WLANs,\n"
```

File: pyntc/devices/aireos_device.py [hunk 13]
```diff
@@ -783,6 +870,7 @@ def enabled_wlans(self):  # noqa: D403
             >>>
         """
         enabled_wlans = [wlan_id for wlan_id, wlan_data in self.wlans.items() if wlan_data["status"] == "Enabled"]
+        log.info("Host %s: List of enabled WLAN IDs: {enabled_wlans}", self.host, enabled_wlans)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:873]:
```suggestion
        log.info("Host %s: List of enabled WLAN IDs: %s", self.host, enabled_wlans)
```

File: pyntc/devices/aireos_device.py [hunk 14]
```diff
@@ -859,20 +947,35 @@ def file_copy(
                 ]
             )
         except CommandListError as error:
+            log.error(
+                "Host %s: File transfer error {FileTransferError.default_message}\n\n{error.message}",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:951]:
```suggestion
                "Host %s: File transfer error %s\n\n%s",
```

File: pyntc/devices/aireos_device.py [hunk 15]
```diff
@@ -859,20 +947,35 @@ def file_copy(
...
+            )
             raise FileTransferError(error.message)
 
         try:
             response = self.native.send_command_timing("transfer download start")
             if "Are you sure you want to start? (y/N)" in response:
                 response = self.show("y", auto_find_prompt=False, delay_factor=delay_factor)
         except CommandError as error:
+            log.error(
+                "Host %s: File transfer error {FileTransferError.default_message}\n\n{error.message}",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:964]:
```suggestion
                "Host %s: File transfer error %s\n\n%s",
```

File: pyntc/devices/aireos_device.py [hunk 16]
```diff
@@ -859,20 +947,35 @@ def file_copy(
...
+                error.message,
+            )
             raise FileTransferError(message=f"{FileTransferError.default_message}\n\n{error.message}")
         except:  # noqa E722
+            log.error("Host %s: File transfer error %s", self.host, FileTransferError)
             raise FileTransferError
 
         if "File transfer is successful" not in response:
-            raise FileTransferError(message=f"Did not find expected success message in response, found:\n{response}")
+            log.error("Host %s: Did not find expected success message in response, found:\n%s", self.host, response)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:975]:
```suggestion
            log.error("Host %s: Did not find expected success message in file transfer response, found:\n%s", self.host, response)
```

File: pyntc/devices/aireos_device.py [hunk 17]
```diff
@@ -938,14 +1041,25 @@ def install_os(self, image_name, controller="both", save_config=True, disable_wl
...
+                    "Host %s: Peer failed to form error for image %s and peer redundancy %s",
+                    self.host,
+                    image_name,
+                    peer_redundancy,
+                )
+                raise OSInstallError(
+                    hostname=f"Host {self.host}: {self.host}-standby", desired_boot=f"{image_name}-{peer_redundancy}"
+                )
 
+            log.info("Host %s: OS image %s installed successfully.", self.host, image_name)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1059]:
```suggestion
            log.info("Host %s: OS image %s installed.", self.host, image_name)
```

File: pyntc/devices/aireos_device.py [hunk 18]
```diff
@@ -1105,6 +1227,7 @@ def redundancy_mode(self):
         """
         ha = self.show("show redundancy summary")
         ha_mode = re.search(r"^\s*Redundancy\s+Mode\s*=\s*(.+?)\s*$", ha, re.M)
+        log.debug("Host %s: Redundancy mode: {ha_mode.group(1).lower()}", self.host, ha_mode.group(1).lower())
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1230]:
```suggestion
        log.debug("Host %s: Redundancy mode: %s", self.host, ha_mode.group(1).lower())
```

File: pyntc/devices/aireos_device.py [hunk 19]
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
                 command_responses.append(command_response)
                 self._check_command_output_for_errors(cmd, command_response)
         except TypeError as err:
+            log.error("Host %s: Netmiko Driver's {err.args[0]}", self.host, err.args[0])
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1388]:
```suggestion
            log.error("Host %s: Netmiko Driver's %s", self.host, err.args[0])
```

File: pyntc/devices/aireos_device.py [hunk 20]
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
...
             raise TypeError(f"Netmiko Driver's {err.args[0]}")
         # TODO: Remove this when deprecating config_list method
         except CommandError as err:
             if not original_command_is_str:
+                log.error(
+                    "Host %s: Command error for commands {entered_commands} with message {err.cli_error_msg}.",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1394]:
```suggestion
                    "Host %s: Command error for commands %s with message %s.",
```

File: pyntc/devices/aireos_device.py [hunk 21]
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
...
+                    self.host,
+                    entered_commands,
+                    err.cli_error_msg,
+                )
                 raise CommandListError(entered_commands, cmd, err.cli_error_msg)
             else:
+                log.error(
+                    "Host %s: Command error for commands {entered_commands} with message %s.",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1402]:
```suggestion
                    "Host %s: Command error for commands %s with message %s.",
```

File: pyntc/devices/aireos_device.py [hunk 22]
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
...
+                    err,
+                )
                 raise err
 
         # TODO: Remove this when deprecating config_list method
         if original_command_is_str:
             return command_responses[0]
 
+        log.debug(
+            "Host %s: Successfully executed command 'show' with responses {command_responses}.",
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1414]:
```suggestion
            "Host %s: Successfully executed show commands with responses %s.",
```

File: pyntc/devices/aireos_device.py [hunk 23]
```diff
@@ -1382,8 +1530,10 @@ def transfer_image_to_ap(self, image, timeout=None):
             time.sleep(1)
 
         if not self._ap_images_match_expected("primary", image):
-            raise FileTransferError(f"Unable to set all APs to use {image}")
+            log.error("Host %s: Unable to set all APs to use %s", self.host, image)
+            raise FileTransferError
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1534]:
```suggestion
            raise FileTransferError(f"Unable to set all APs to use {image}")
```

File: pyntc/devices/aireos_device.py [hunk 24]
```diff
@@ -1404,6 +1554,7 @@ def uptime(self):
         hours += days * 24
         minutes += hours * 60
         seconds = minutes * 60
+        log.debug("Host %s: Device has been up for %s seconds", self.host, seconds)
```
Comment (jmcgill298) [inline pyntc/devices/aireos_device.py:1557]:
```suggestion
        log.debug("Host %s: Device has been up for %d seconds", self.host, seconds)
```

**Review by jmcgill298 — CHANGES_REQUESTED** [review]
@mundruid I started with these, so far I have only made it through the aireos driver

**Comment by mundruid** [comment]
> @mundruid I started with these, so far I have only made it through the aireos driver

@jmcgill298 Thanks for the review!

**Comment by jeffkala** [comment]
@jmcgill298 @mundruid I know its a longshot being this PR is so old, but was this ready to go? Worth revisiting to get it over the finish line or just close?

**Comment by mundruid** [comment]
@jeffkala On my part, this was ready after addressing Ken's comment about the not using `f-strings` in all the logs. I have added the tests and all the logging. However, It is a long one to review and I think that is why we have not merged it yet.

**Comment by jmcgill298** [comment]
@jeffkala I put several comments above where things do not look quite right. I did not get through all of the updates, but I suspect similar comments to be repeated

**Comment by jeffkala** [comment]
@mundruid @jmcgill298 thanks, will try and work on getting the rest updated and get this PR over the finish line.

**Comment by jeffkala** [comment]
taking it over the finish line in this pr #271

**Pull request closed without merging**